### PR TITLE
Simplify __getitem__ and __setitem__

### PR DESCRIPTION
**Pull request by okuta**
~Please merge #747 first.~
This PR refactor `__getitem__` and `__setitem__` .
These have same code. So, I create new helper functions.
And, this PR contains some small  fixing.

### CODE REVIEW

**Comment by unnonouno** [comment]
Can you split this PR?

**Comment by okuta** [comment]
I split PR.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -3060,162 +3085,67 @@ cpdef _scatter_op_mask_single(ndarray a, ndarray mask, v, Py_ssize_t axis, op):
...
             _prepare_multiple_array_indexing(a, adv_slices)
-        _scatter_op_single(a_interm, reduced_idx, value, li=li, ri=ri, op=op)
+        _scatter_op_single(a_interm, reduced_idx, value, li, ri, op)
         return
 
+    y = _simple_getitem(a, slice_list)
     if op == 'update':
-        v = a[tuple(slices)]
-        if isinstance(value, ndarray):
-            y, x = broadcast(v, value).values
```
Comment (unnonouno) [inline cupy/core/core.pyx]: This broadcast is removed in the PR. Is it OK?

File: cupy/core/internal.pxd [hunk 2]
```diff
@@ -27,3 +27,4 @@ cpdef vector.vector[Py_ssize_t] infer_unknown_dimension(
     vector.vector[Py_ssize_t]& shape, Py_ssize_t size) except *
 
 cpdef slice complete_slice(slice slc, Py_ssize_t dim)
+cpdef tuple complete_slice_list(list slice_list, Py_ssize_t ndim)
```
[the remark this answers is not in the document]
Comment (okuta) [reply cupy/core/internal.pxd]: OK

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -2781,6 +2637,175 @@ right_shift = _create_bit_op(
...
+
+    if not mask_exists and len(slice_list) > ndim + n_newaxes:
+        raise IndexError('too many indices for array')
+    return slice_list, advanced, mask_exists
+
+
+cdef Py_ssize_t _get_mask_index(list slice_list) except *:
+    cdef Py_ssize_t i, n_not_slice_none, mask_i
+    n_not_slice_none = 0
+    mask_i = -1
```
[the remark this answers is not in the document]
Comment (okuta) [reply cupy/core/core.pyx]: Yes.

File: cupy/core/core.pyx (hunk 1 above)
Comment (okuta) [reply cupy/core/core.pyx]: Yes, the array is broadcated by elementwise kernel.

File: cupy/core/internal.pxd (hunk 2 above)
[the remark this answers is not in the document]
Comment (unnonouno) [reply cupy/core/internal.pxd]: I mean end of the file.

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -3060,162 +3085,67 @@ cpdef _scatter_op_mask_single(ndarray a, ndarray mask, v, Py_ssize_t axis, op):
...
+        _scatter_update_mask_kernel(src, mask, mask_scanned, a)
     elif op == 'add':
-        _scatter_add_mask_kernel(v, mask, mask_scanned, a)
+        _scatter_add_mask_kernel(src, mask, mask_scanned, a)
     else:
         raise ValueError('provided op is not supported')
 
 
 cpdef _scatter_op(ndarray a, slices, value, op):
-    cdef Py_ssize_t i, ndim, n_newaxes, n_ellipses, ellipsis, axis
```
[the remark this answers is not in the document]
Comment (okuta) [reply cupy/core/core.pyx]: Yes.

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -2781,6 +2637,175 @@ right_shift = _create_bit_op(
...
+            else:
+                dim = (s_stop - s_start + 1) // s_step + 1
+
+            if dim == 0:
+                strides.push_back(a._strides[j])
+            else:
+                strides.push_back(a._strides[j] * s_step)
+
+            if s_start > 0:
+                offset += a._strides[j] * s_start
```
[the remark this answers is not in the document]
Comment (okuta) [reply cupy/core/core.pyx]: It is checked by `if s_start > 0:`.

File: cupy/core/internal.pxd (hunk 2 above)
[the remark this answers is not in the document]
Comment (okuta) [reply cupy/core/internal.pxd]: I tried it, but, flake8 was failed.

**Comment by okuta** [comment]
I fixed.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**